src/lsm/compaction.zig

```diff
@@ -311,293 +323,234 @@ pub fn CompactionType(
-                    });
-                    swap_buffers(&compaction.filter, &compaction.table_builder.filter_block);
-                    assert(compaction.filter.ready);
-                }
-            }
+            // Finalize the filter block if it's full or if it contains pending data blocks
+            // when there's no more merged values to fill them.
+            if (compaction.table_builder.filter_block_full() or 
+                (merge_iterator.empty() and !compaction.table_builder.filter_block_empty())
+            ) {
+                compaction.table_builder.filter_block_finish(.{
+                    .cluster = compaction.grid.superblock.working.cluster,
+                    .address = compaction.grid.acquire(),
+                });
 
-            if (compaction.table_builder.index_block_full() or
-                compaction.merge_iterator.empty())
-            {
-                if (compaction.table_builder.index_block_empty()) {
-                    assert(compaction.drop_tombstones);
-                    assert(tombstones_dropped > 0);
-                } else {
-                    const snapshot_min = compaction.snapshot;
-                    const table = compaction.table_builder.index_block_finish(.{
-                        .cluster = compaction.grid.superblock.working.cluster,
-                        .address = compaction.grid.acquire(),
-                        .snapshot_min = snapshot_min,
-                    });
-                    compaction.queue_manifest_update(&compaction.insert_level_b, &table);
-
-                    swap_buffers(&compaction.index, &compaction.table_builder.index_block);
-                    assert(compaction.index.ready);
-                }
+                // Mark the finished filter block as writable for the next compact_tick() call.
+                compaction.filter.block = compaction.table_builder.filter_block;
+                assert(!compaction.filter.writable);
+                compaction.filter.writable = true;
             }
-        }
-
-        fn iterator_a_callback(iterator_a: *IteratorA) void {
-            const compaction = @fieldParentPtr(Compaction, "iterator_a", iterator_a);
-            compaction.io_callback();
-        }
-
-        fn iterator_b_callback(iterator_b: *IteratorB) void {
-            const compaction = @fieldParentPtr(Compaction, "iterator_b", iterator_b);
-            compaction.io_callback();
-        }
 
-        fn io_callback(compaction: *Compaction) void {
-            compaction.io_pending -= 1;
-            if (compaction.io_pending == 0) compaction.tick_done();
-        }
+            // Finalize the index block if it's full or if it contains pending data blocks
+            // when there's no more merged values to fill them.
+            if (compaction.table_builder.index_block_full() or
+                (merge_iterator.empty() and !compaction.table_builder.index_block_empty())
+            ) {
```

Comment:
:+1: Perfect.

Fwiw I think our formatting convention is:
```suggestion
            if (compaction.table_builder.index_block_full() or
                (merge_iterator.empty() and !compaction.table_builder.index_block_empty()))
            {
```